### ISSUE TITLE
Default export of the module has or is using private name 'Props'

### ISSUE THREAD

**Issue by ly-sniper** [vuejs/language-tools#1232]
when upgrade volar to ^0.34.9 ,  there is a ts warning `Default export of the module has or is using private name 'Props'` at the top of  template

```js
//error info -> Default export of the module has or is using private name 'Props'. ts(4082)
<template>test</template>
<script lang="ts" setup>
  //step1: declear type
  interface Props {
    foo: string;
  }
 // setp2: defineProps
  defineProps<Props>(); 
</script>
```

but if definePorps like below, the warning disappear
```js
<template>test</template>
<script lang="ts" setup>
  defineProps<{
    foo: string;
  }>();
</script>
```

**Comment by johnsoncodehk** [comment]
Hi @hayabusa2009, try remove `"declaration": true` in tsconfig or move interface Props to script block.

```html
<script lang="ts">
//step1: declear type
interface Props {
  foo: string;
}
</script>

<script lang="ts" setup>
// setp2: defineProps
defineProps<Props>(); 
</script>
```

**Comment by mobayen** [comment]
Why Volar is complaining about the interface anyway?
Is it a bad practice?

**Comment by johnsoncodehk** [comment]
@mobayen You just need move interface from `<script setup>` to `<script>` if it is a part of emit type, if in `<script setup>`, the interface is actually inside `setup()` scope so it can't be emit.

```ts
...
setup() {
	interface Props {
      foo: string;
    }
}
```

**Comment by hayabusa2009** [comment]
> Hi @hayabusa2009, try remove `"declaration": true` in tsconfig or move interface Props to script block.
> 
> ```
> <script lang="ts">
> //step1: declear type
> interface Props {
>   foo: string;
> }
> </script>
> 
> <script lang="ts" setup>
> // setp2: defineProps
> defineProps<Props>(); 
> </script>
> ```

@johnsoncodehk thank you for your replay and I try your advices but failed.
I want to know is there a way to **_extends_** thrid-part library component types in the setup block?
for example
```js
<script lang="ts" setup>
  import type { ModalProps } from 'third-part-library/Modal';
  interface Props extends ModalProps {
    visible: boolean;
  }
  const props = defineProps<Props>();

</script>
```

**Comment by johnsoncodehk** [comment]
@hayabusa2009 Have you try move `interface Props` to `<script>` tag? It it not works please provide a minimal reproduction.

```
<script lang="ts">
  import type { ModalProps } from 'third-part-library/Modal';
  interface Props extends ModalProps {
    visible: boolean;
  }
</script>

<script lang="ts" setup>
  const props = defineProps<Props>();
</script>
```

**Comment by hayabusa2009** [comment]
@johnsoncodehk  when remove `composite:true` from tsconfig.json, warning disappear. tks again!

**Comment by zengxinhandsome** [comment]
> @johnsoncodehk when remove `composite:true` from tsconfig.json, warning disappear. tks again!

works for me, thx

**Comment by caio-2k** [comment]
> @hayabusa2009 Have you try move `interface Props` to `<script>` tag? If it not works please provide a minimal reproduction.
> 
> ```
> <script lang="ts">
>   import type { ModalProps } from 'third-part-library/Modal';
>   interface Props extends ModalProps {
>     visible: boolean;
>   }
> </script>
> 
> <script lang="ts" setup>
>   const props = defineProps<Props>();
> </script>
> ```

Life savier, ty very much!

**Comment by iamgabrielsoft** [comment]
removing declaration from tsconfig.json file works for me

**Comment by vaban-ru** [comment]
Hello, I have same issue when try to build in lib mode

**Comment by iamgabrielsoft** [comment]
Can you provide a minimal reproduction code?

**Comment by vaban-ru** [comment]
I will try to describe the problem in more detail. I have a small component library. Vue 3 + TS + Vite.
In package.json I have this version of libraries:
```
"vue": "^3.2.37",
"vue-tsc": "^0.40.13"
"vite": "^3.1.0",
"typescript": "^4.8.3",
```

In tsconfig.json
```
{
  "compilerOptions": {
    "target": "esnext",
    "useDefineForClassFields": true,
    "module": "esnext",
    "moduleResolution": "node",
    "strict": true,
    "jsx": "preserve",
    "sourceMap": true,
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "lib": ["esnext", "dom"],
    "declaration": true,
    "emitDeclarationOnly": true,
    "declarationDir": "./dist/types",
    "isolatedModules": true,
    "skipLibCheck": true,
    "types": [
      "vite/client"
    ],
    "baseUrl": ".",
    "paths": {
      "~/*": ["./*"],
      "@/*": ["./src/*"]
    }
  },
  "include": [
    "src/index.ts",
    "src/components/**/*.ts",
    "src/components/**/*.vue",
    "src/composables/**/*.ts",
    "src/mocks/**/*.ts",
    "src/env.d.ts"
  ]
}
```

In vite.config.ts
```
const path = require("path");
const { defineConfig } = require("vite");
import vue from "@vitejs/plugin-vue";

module.exports = defineConfig(() => {
  const rootPath = path.resolve(process.cwd());
  const srcPath = `${rootPath}/src`;
  return {
    resolve: {
      alias: {
        "~": rootPath,
        "@": srcPath,
      },
    },
    plugins: [vue()], 
    build: {
      lib: {
        entry: path.resolve(__dirname, "src/index.ts"),
        name: "UiKitNext",
      },
      rollupOptions: {
        external: [
          "vue",
        ], 
        output: {
          exports: "named",
          globals: {
            vue: "Vue",
          },
        },
      },
      emptyOutDir: false,
    },
  };
});
```

Example of component:
```vue
<template>
  <div v-show="active" class="tab">
    <slot />
  </div>
</template>

<script lang="ts">
import { IEvent } from "@/components/types";

export interface ISvTabProps {
  title: string;
  id: string;
  actions?: IEvent[];
  hidden?: boolean;
  parentId?: number;
}
</script>

<script setup lang="ts">
import { ComponentInternalInstance, getCurrentInstance, ref } from "vue";

const props = withDefaults(defineProps<ISvTabProps>(), {
  hidden: false,
});

const active = ref(true);
let parentId: number | undefined;
const app = getCurrentInstance();
const $bus = app?.appContext.config.globalProperties.$bus;

const searchParentID = (
  parent: ComponentInternalInstance | undefined | null
) => {
  if (parent) {
    if (parent.type.__name === "SvTabs") {
      parentId = parent.uid;
      return;
    } else {
      searchParentID(parent.parent);
    }
  }
  return;
};

searchParentID(app?.parent);

$bus.on("setActiveTab", (tab: ISvTabProps) => {
  if (tab.parentId === parentId && tab.id === props.id) {
    active.value = true;
  }
  if (tab.parentId === parentId && tab.id !== props.id) {
    active.value = false;
  }
});

$bus.fire("createTab", {
  id: props.id,
  title: props.title,
  parentId,
  actions: props.actions,
  hidden: props.hidden,
});
</script>
```

Build script in package.json
```
"build": "rimraf dist && vue-tsc && vite build",
```

When I run the vue-tsc, everything works fine and the `dist/types` are generated.

When I run the build script I get the following error:

```
Unexpected "="
29 |    setup(__props: any) {
30 |
31 |  const props = __props as { t props = wit, aults(defi, ops<ISvTabProps>(), hiddenn: false,,
   |                                     ^
32 |
33 |  const active =  }

transforming (131) src\components\Widgets\SvDepositForm\SvDepositForm.vueerror during build:
Error: Transform failed with 1 error:
D:/dev/norvikbank-next/ui-kit-next/src/components/Base/SvTab/SvTab.vue?vue&type=script&setup=true&lang.ts:31:35: ERROR: Unexpected "="
```

**Comment by chesterlaykin** [comment]
My error had to do with Vitest: if I choose vitest when installing (`npm init vue@latest`) , the tsconfig is different and has some files included with "references":
```
{
  "files": [],
  "references": [
    {
      "path": "./tsconfig.config.json"
    },
    {
      "path": "./tsconfig.app.json"
    },
    {
      "path": "./tsconfig.vitest.json"
    }
  ]
}
```
I tried to change stuff in those three files, but didn't manage to get rid of the errors.
If I remove the references array and add this from the original vue tsconfig the errors disappear:

```
  "include": ["env.d.ts", "src/**/*", "src/**/*.vue"],
  "files": [],
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    }
  }
```
Here is a [reproduction](https://github.com/chesterlaykin/vue3-ts-volar-vitest-issue)

**Comment by vaban-ru** [comment]
Create some [reproduction repo ](https://github.com/vaban-ru/vue-tsc-reproduction-issue)

**Comment by vaban-ru** [comment]
Maybe all the same it's not in the library but in Vue itself 
https://github.com/vitejs/vite/issues/9836
https://github.com/vuejs/core/issues/5830

**Comment by StepanMynarik** [comment]
Same problem here. I discovered it when using the new reactivity transform feature.
Anyway, this error occurs when using interface with defineProps in general I found out.

Can be solved by exporting the interface, but that doesn't seem right to me since no export is used in Vue docs in this use case.

**Comment by johnsoncodehk** [comment]
> Create some [reproduction repo ](https://github.com/vaban-ru/vue-tsc-reproduction-issue)

The error is throw by `vite build` not `vue-tsc`, please consider reporting it to vite repo.

**Comment by johnsoncodehk** [comment]
> My error had to do with Vitest: if I choose vitest when installing (`npm init vue@latest`) , the tsconfig is different and has some files included with "references":

I ca only got `Cannot find module '../HelloWorld.vue' or its corresponding type declarations.` error in the repo:

```
$ npm run type-check

> vue3-ts-volar-vitest-issue@0.0.0 type-check
> vue-tsc --noEmit -p tsconfig.vitest.json --composite false

src/components/__tests__/HelloWorld.spec.ts:4:24 - error TS2307: Cannot find module '../HelloWorld.vue' or its corresponding type declarations.

4 import HelloWorld from '../HelloWorld.vue'
                         ~~~~~~~~~~~~~~~~~~~


Found 1 error in src/components/__tests__/HelloWorld.spec.ts:4
```

**Comment by johnsoncodehk** [comment]
@StepanMynarik could you provide repro case?

**Comment by SavkaTaras** [comment]
The closes you can get, it to use it in this kind of format:

```
<script setup lang="ts">
export interface Props {
  name?: string;
};

const props = withDefaults(defineProps<Props>(), {
  name: '',
});

const emit = defineEmits<{
  (e: 'custom-event', id: number): void;
  (e: 'custom-update'): void;
}>();

const emitEvent = () => {
  emit('custom-event', 2);
};

const emitUpdate = () => {
  emit('custom-update');
};
</script>
```

**Comment by ffxsam** [comment]
Just ran into this today. Exporting the interface got rid of the error, but this is not a good developer experience.

**Comment by vknowles-rv** [comment]
Why is this closed? Based on how Vue 3 outlines how to use `<script setup>`, volar is not working properly. 

Littering a project with `<script>` as a work around is exhausting, cruft, and confusing in medium/large projects.

`export interface` works, but isn't necessary in other IDEs. Not to mention, if anything that goes above it that isn't an import breaks the workaround.

Apologies for coming off rude, but workarounds doesn't constitute closing issues.

**Comment by rezelute** [comment]
I agree with this, the plugin isnt working how the docs say it should work ..?

**Comment by panzerdp** [comment]
A solution is to use to annotate `Props` using `type` instead of `interface`:

```vue
<script setup lang="ts">
type Props = {
  names: string[]
  value: number
}
const props = defineProps<Props>()

//...
</script>
```

Works in Volar `1.0.24`.

**Comment by Ge6ben** [comment]
Same here!

**Comment by Ge6ben** [comment]
> Why is this closed? Based on how Vue 3 outlines how to use `<script setup>`, volar is not working properly.
> 
> Littering a project with `<script>` as a work around is exhausting, cruft, and confusing in medium/large projects.
> 
> `export interface` works, but isn't necessary in other IDEs. Not to mention, if any code inserted above `export interface` that isn't an import breaks the workaround.
> 
> Apologies for coming off rude, but workarounds doesn't constitute closing an issue.

Same here!

**Comment by so1ve** [comment]
Hey guys - for those who subscribes this issue - this is likely a limitation in Typescript. `interface` can be extended, but `type aliases` can't:

```ts
interface Foo {
  foo: 1
}
// No error! Foo has been extended
interface Foo {
  bar: 1
}
// -------
type Bar = {}
// Oops, error!
type Bar = {}
```
Why `export interface` works? Volar will hoist all export statements to top-level, so they can be imported from other files. However, though hoisting all interfaces seems to be a viable solution, it's un unacceptable (if the interface references some type alias defined in `script setup`, what should we do?).

In looking for a doable solution and I can implement it in a PR. Any help is appreciated!

**Comment by amalitsky** [comment]
Please, reopen this issue.

Having it closed sends the wrong message, and motivates people folks to open new issues for the same topic.

My issue is the same as described above: I can't define `props` interface extending a third-party library interface.
It's in the _non-setup_ `script` tag.

I don't seem to be able to ignore this error with `/* @vue-ignore */` either.

And this is a library of components.

**Comment by frankypixels** [comment]
It really doesn't make the best sense to close the issue if the docs say one thing and the plugin faults the same thing. It's very confusing to follow what you think is convention then get dinged for it and use work arounds.

**Comment by so1ve** [comment]
@frankypixels Any reproduction? I can't reproduce this issue anymore.

**Comment by frankypixels** [comment]
This is exactly that I had yesteday 

`
<script setup lang="ts">

  interface Notes {
    id: string
    content: string
  }
  
  const notes = defineProps<Notes>()

</script>
`

**Comment by Grenek** [comment]
It worked for me after updating to pre-release version 1.8.23

**Comment by johnsoncodehk** [comment]
Starting from a certain version, when there is only one script setup block, the setup code will be generated to the file root instead of the function scope to avoid this problem.

**Comment by mityaua** [comment]
2.1.4 - still actual. But with defineComponent syntax.

**Comment by charnog** [comment]
I created a repository to demonstrate the issue where a `<script setup>` component (with props defined via `interface Props`) is used within a `defineComponent`, causing a “Default export…” error.

Repository: https://github.com/charnog/vue-default-export

We initially started our project using only the `defineComponent` syntax (Vue 2). When Vue 3 was released, we began using the Composition API for all new components. Naturally, we use the “old” `defineComponent` components within the new ones. However, after some package updates, this error started appearing.

This might also be related to @mityaua’s issue, but I’m not entirely sure.

Let me know if you’d like me to create a distinct issue for this.

**Comment by charnog** [comment]
@webdesignberlin Yep, thank you! This “workaround” was already described [here](https://github.com/vuejs/language-tools/issues/1232#issuecomment-1165416439). There’s also the `type Props` workaround, or you can use `defineProps<{ msg: string }>` instead. The [official docs](https://vuejs.org/guide/typescript/composition-api.html#typing-component-props) explain how to type `Props` in a more formal way without requiring an export.

My question isn’t about “how to fix it in a component” but rather “what is the best approach and is it possible to address this in volar/vue-tsc?” If not, then what should be considered the official approach?

**Comment by KazariEX** [comment]
This is a limitation of TS.

```ts
// setup.ts
import { defineComponent } from "vue";

interface Props {
  msg: string;
}

export default defineComponent({
  __typeProps: {} as Props
});
```

```ts
// define.ts
import { defineComponent } from "vue";
import Setup from "./setup";

export default defineComponent({
  components: {
    Setup,
  },
});
```

The currently discovered solution that can be implemented in this repository is:

```ts
export default defineComponent({
  __typeProps: {
    ...{} as Props
  }
});
```

**Comment by KazariEX** [comment]
@charnog please feel free to open a new issue.

**Comment by charnog** [comment]
@KazariEX, I just want to keep you informed.

The problem had been getting worse and worse. After I updated TypeScript to 5.6.3, the number of errors became unmanageable. TypeScript 5.7.3 introduced even more errors, including strange ones like “there is an interface in the node_modules package that is not exported.”

Then, in the [3.13.0](https://github.com/vuejs/create-vue/releases/tag/v3.13.0) release of [create-vue](https://github.com/vuejs/create-vue), `composite: true` was removed in [this pull request](https://github.com/vuejs/create-vue/pull/635). Removing `composite: true` from our configs resolved all the issues (including the strange ones) within our project. So, there is no more need to fix this (at least for now).

I believe the problem is related to [declaration](https://www.typescriptlang.org/tsconfig/#declaration), which is forced to true when composite: true is enabled. As a result, TypeScript treats the app as a public library and expects every interface to be “public.”
